mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -17,117 +16,114 @@ impl S3CrtClient {
-            .client
-            .new_request_template("PUT", &self.bucket)
+            .inner
+            .new_request_template("PUT", bucket)
             .map_err(S3RequestError::construction_failure)?;
 
-        let key = format!("/{}", self.key);
+        let key = format!("/{}", key);
         message
             .set_request_path(&key)
             .map_err(S3RequestError::construction_failure)?;
 
-        if self.params.trailing_checksums {
+        if params.trailing_checksums {
             let checksum_config = ChecksumConfig::trailing_crc32c();
             message.set_checksum_config(Some(checksum_config));
         }
 
-        let (body_async_stream, writer) = async_stream::new_stream(&self.client.allocator);
+        let (body_async_stream, writer) = async_stream::new_stream(&self.inner.allocator);
         message.set_body_stream(Some(body_async_stream));
 
-        if is_empty {
-            message
-                .add_header(&Header::new("Content-Length", 0.to_string()))
-                .map_err(S3RequestError::construction_failure)?;
-        }
+        span.in_scope(|| debug!(?bucket, ?key, ?params, "make request"));
+
+        let review_callback = ReviewCallbackBox::default();
+        let callback = review_callback.clone();
 
-        self.span
-            .in_scope(|| debug!(bucket=?self.bucket, ?key, params=?self.params, "make request"));
-        let body =
-            self.client
-                .make_simple_http_request(message, MetaRequestType::PutObject, self.span.clone(), |result| {
-                    ObjectClientError::ClientError(S3RequestError::ResponseError(result))
-                })?;
-        Ok((body, writer))
+        let mut options = S3CrtClientInner::new_meta_request_options(message, MetaRequestType::PutObject);
+        options.on_upload_review(move |review| callback.invoke(review));
+        let body = self
+            .inner
+            .make_simple_http_request_from_options(options, span, |result| {
+                ObjectClientError::ClientError(S3RequestError::ResponseError(result))
+            })?;
+
+        Ok(S3PutObjectRequest {
+            body,
+            writer,
+            review_callback,
+        })
     }
 }
 
+type ReviewCallback = dyn FnOnce(UploadReview) -> bool + Send;
+
+#[derive(Clone, Default)]
+struct ReviewCallbackBox {
+    callback: Arc<Mutex<Option<Box<ReviewCallback>>>>,
+}
+
+impl std::fmt::Debug for ReviewCallbackBox {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ReviewCallbackBox").finish()
+    }
+}
+
+impl ReviewCallbackBox {
+    fn set(&mut self, callback: impl FnOnce(UploadReview) -> bool + Send + 'static) {
+        let _ = self.callback.lock().unwrap().insert(Box::new(callback));
+    }
+
+    fn invoke(self, review: UploadReview) -> Result<(), i32> {
+        let mut callback = self.callback.lock().unwrap();
+        if let Some(callback) = callback.take() {
```

Comment:
Somewhere it would be a nice sanity check to make sure this callback is invoked at most once, and maybe return `Err` on a second call. I don't know if the right place to do that check is here or in the CRT binding -- probably in the binding since what we're really checking is that the CRT is holding up its end of the contract?